postgresql/helpers.go

```diff
@@ -92,6 +92,45 @@ func revokeRoleMembership(db QueryAble, role, member string) error {
 	return nil
 }
 
+// withRolesGranted temporary grants, if needed, the roles specified to connected user
+// (i.e.: the admin configure in the provider) and revoke them as soon as the
+// callback func has finished.
+func withRolesGranted(txn *sql.Tx, roles []string, fn func() error) error {
+	// Get current user
+	var currentUser string
+	err := txn.QueryRow("SELECT CURRENT_USER").Scan(&currentUser)
+	switch {
+	case err == sql.ErrNoRows:
+		return fmt.Errorf("SELECT CURRENT_USER returns now row, this is quite disturbing")
+	case err != nil:
+		return errwrap.Wrapf("error while looking for the current user: {{err}}", err)
+	}
+
+	var granted []string
+	for _, role := range roles {
+		roleGranted, err := grantRoleMembership(txn, role, currentUser)
+		if err != nil {
+			return err
+		}
+		if roleGranted {
+			granted = append(granted, role)
+		}
+	}
+
+	if err := fn(); err != nil {
+		return err
+	}
+
+	for _, role := range granted {
+		err := revokeRoleMembership(txn, role, currentUser)
```

Comment:
One problem I could see is that you fail to revoke a role and return an error, leaving the other roles intact, but letting the caller know only about the first revocation that failed. You could use something like [gojek/valkyrie](https://github.com/gojek/valkyrie), or collect the errors yourself, instead.

Reply:
Not sure to get this point.
It's executed in a transaction so if this returns an error,  everything will be rollback so the other roles will no be left "intact"

Reply:
RIght